### PR TITLE
Final Spec 4 to be merged

### PR DESCRIPTION
**Pull request by konnov**
Since there was plenty of experimental work done in #51, I have chosen the commit a6ac70638ebcc3b88193d52496e1fd8ce1fa0b48 to contain the final results for `Spec 4`. When we commit this PR, we will close #51.

The further works that was done in the branch igor/rethink-blocks-split will go into `Spec4b`.

### CODE REVIEW

File: spec-abstract-noclosure/ffg.tla [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    /\ IsValidCheckpoint(vote.source)
+    /\ IsValidCheckpoint(vote.target)
+    /\ vote.source[2] < vote.target[2]
+    /\ IsLeftAncestorOfRight(vote.source[1], vote.target[1])
+    (* See ../spec/ffg.tla:
+    /\ has_block_hash(vote.message.ffg_source.block_hash, node_state)
+    /\ get_block_from_hash(vote.message.ffg_source.block_hash, node_state).slot = vote.message.ffg_source.block_slot
+    /\ has_block_hash(vote.message.ffg_target.block_hash, node_state)
+    /\ get_block_from_hash(vote.message.ffg_target.block_hash, node_state).slot = vote.message.ffg_target.block_slot
+     *)
```
Comment (Kukovec) [inline spec-abstract-noclosure/ffg.tla:134]:
```suggestion
```

File: spec-abstract-noclosure/ffg.tla [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+\* @type: ($block, $block) => Bool;
+IsLeftAncestorOfRight(before, after) ==
+   /\ before.slot <= after.slot
+   /\ \/ before \in chain1 /\ after \in chain1
+      \/ before \in chain2 /\ after \in chain2
+   (*
+   /\ \/ before.body >= 0 /\ after.body >= 0 /\ before.body <= after.body
+      \/ before.body < 0 /\ after.body < 0 /\ -before.body <= -after.body
+      \/ before.body >= 0 /\ after.body < 0 /\ before.body <= -after.body /\ after.body <= chain2_fork_block_number
+    *)
```
Comment (Kukovec) [inline spec-abstract-noclosure/ffg.tla:101]:
```suggestion
```

**Review by Kukovec — APPROVED** [review]
LGTM, I assume these are well-tested anyhow.

File: spec-abstract-noclosure/ffg.tla [hunk 3]
```diff
@@ -0,0 +1,303 @@
+----------------------------- MODULE ffg -----------------------------
+(*
+ * TODO
```
Comment (thpani) [inline spec-abstract-noclosure/ffg.tla:3]: TODO?

**Review by thpani — APPROVED** [review]
lgtm

**Review by thpani — COMMENTED** [review]
Shall we add a README though?

**Pull request merged**